_posts/2018-04-11-sign-to-contract.md

```diff
@@ -0,0 +1,400 @@
+- conclusions
+
+# A bit of history
+When hearing the name *sign-to-contract*, 
+you may wonder from where the contract comes out.
+To find an answer we have to go back (in 2012) to one of the [first applications](https://arxiv.org/pdf/1212.3257.pdf) that was developed: *pay-to-contract* (*sign-to-contract* elder brother) which basically allows to embed a contract inside a receiver public key.
+When a customers performs a payment to a merchant, with *pay-to-contract*, 
+he can bind the bill containing the details of the purchase to the receiving address.
+
+Although other uses of these techniques are possible, 
+we focus on timestamping purposes 
+and we take as a reference the [PR](https://github.com/opentimestamps/python-opentimestamps/pull/14) to OpenTimestamps 
+(and the preceding [issue](https://github.com/opentimestamps/python-opentimestamps/issues/12)) by Andrew Poelstra.
+
+# A bit of math: elliptic curve commitments
+Now we need to get our hands into cryptography to unveil how this math trick works under the hood
+(don't get scared, we provide a quick recap of the results at the end of the section).
+
+Given an elliptic curve with generator `G` and a second-preimage resistant hash function `h`.
+The map `m,P -> h(P||m)G + P`, where `||` stands for concatenation, `m` is the message to commit and `P` is an elliptic curve point, is a valid commitment operation.
+The output of the map is `C(m,P)`, which is an elliptic curve point.
+
+`m` is the so-called contract and `P` is a public key used for a certain (and possibly independent) purpose.
+The public key `C(m,P)` is a commitment to `m` and `P` 
+in the sense that if the input is changed, so `(m,P)!=(m',P')`, 
+then the output changes, `C(m,P)!=C(m',P')`.
+Thus we can use `C` as a commitment operation (like a hash function) inside a timestamp proof.
+
+Now, to curb the abstraction, choose `secp256k1` as elliptic curve and `sha256` as hash function. 
+The motivation is that we want to use Bitcoin as a notary, 
+and Bitcoin itself grounds its working on the integrity of these two ingredients.
+We won't use directly the corresponding `C`, but instead a slightly modified version: `OpSecp256k1Commitment`.
+It takes one input, `P||m`, and outputs the x-coord of the elliptic curve point `C(m,P)`.
+
+**Recap:** 
+pick a message to commit (also called contract) `m` and a public key `P` (on the curve `secp256k1`), concatenate them (`P||m`), then compute:
+```
+C(m,P) = SHA256(P||m)G + P 
+OpSecp256k1Commitment(P||m) = C(m,P).x
+```
+The value obtained is an elliptic curve point that serves two (independent) purposes: 
+is a public key __and__ is a commitment to `(m,P)` 
+(or only to `m`, if `P` is given before).
+
+To timestamp we need to write the x-coord of the elliptic curve point `C(m,P)` on the blockchain, 
+there are two places where such points can be found, leading to two techniques:
+- Address/Public key: *pay-to-contract*
+- Signature: *sign-to-contract*
+
+## *pay-to-contract*
```

Comment:
It's not clear to me that you actually need this section at all: it's of historical interest, and not relevant as an intro to timestamping with secp256k1 commitments.